wdl-ast/src/v1/task/common/container/value.rs

```diff
@@ -0,0 +1,407 @@
+                        .map(|uri| Value {
+                            kind: Kind::String(uri),
+                            expr,
+                        })
+                        .map_err(|e| Error::ParseString(ParseUriError::Uri(e)));
+                }
+                // An array of literal strings.
+                LiteralExpr::Array(a) => {
+                    let mut all_errors: Vec<ParseUriError> = Vec::new();
+
+                    let (errors, literal_strings): (Vec<_>, Vec<_>) = a
+                        .elements()
+                        .map(|expr| {
+                            expr.clone()
+                                .into_literal()
+                                .and_then(|literal| literal.into_string())
+                                .ok_or(ParseUriError::InvalidExpressionInArray { expr })
+                        })
+                        .partition(std::result::Result::is_err);
+
+                    // SAFETY: we ensured that only results that are [`Err()`] are
+                    // partitioned into this vec, so each of these will always
+                    // unwrap.
+                    all_errors.extend(errors.into_iter().map(std::result::Result::unwrap_err));
+
+                    let (errors, uris): (Vec<_>, Vec<_>) = literal_strings
+                        .into_iter()
+                        // SAFETY: we ensured that only results that are [`Ok()`] are
+                        // partitioned into this vec, so each of these will always
+                        // unwrap.
+                        .map(std::result::Result::unwrap)
+                        .map(Uri::try_from)
+                        .partition(std::result::Result::is_err);
+
+                    // SAFETY: we ensured that only results that are [`Err()`] are
+                    // partitioned into this vec, so each of these will always
+                    // unwrap.
+                    all_errors.extend(
+                        errors
+                            .into_iter()
+                            .map(std::result::Result::unwrap_err)
+                            .map(ParseUriError::Uri),
+                    );
+
+                    if !all_errors.is_empty() {
+                        return Err(Error::ParseArray(all_errors));
+                    }
+
+                    let uris = uris
+                        .into_iter()
+                        // SAFETY: we ensured that only results that are [`Ok()`] are
+                        // partitioned into this vec, so each of these will always
+                        // unwrap.
+                        .map(std::result::Result::unwrap)
+                        .collect();
+
+                    return Ok(Value {
+                        kind: Kind::Array(uris),
+                        expr,
+                    });
+                }
+                _ => {}
+            }
+        }
+
+        Err(Error::InvalidExpression { expr })
```

Comment:
Is this right? I believe that any expression that can be coerced into a `String` or `Array[String]` is valid. Our workflows repo is in the minority by using literal strings for the container. Most other* WDL repos use combinations of placeholders and runtime declarations. I think we can argue that's bad practice, but it should be parsed as valid.

*"most" as defined by my informal survey

Reply:
You're right, but it's going to require type-checking to make that work. So I settled on this for now and think we should change it in the future. If you agree, I'll create an issue to track it.

Reply:
I guess I'm a bit confused about all the plumbing between the pieces here. The class of errors I'm concerned with are expressions with placeholders and name references. I don't see anything in Arena or Gauntlet related to these, but the code _here_ should be firing an error in those cases. Is this error just not exposed anywhere ATM?

There also should be test cases in this file for those styles of expressions, which is maybe contributing to my confusion here.

Reply:
Oh I see what you're concerned about. This does throw an error but, ultimately, that error gets ignored on line 150 and 167 of `wdl-lint/src/rules/container_value.rs` (because I just check if the result is `Ok()`—I don't do anything if it's an `Err`).

Put another way, the current `Value` for this field doesn't respect coercion rules, but that's because coercion is complicated and needs to wait for the bigger coercion PR to be able actually do the coercion.

I've put a test to demonstrate that the use case you mention above is simply ignored.

Reply:
That's great @claymcleod . Answered my question and LGTM 👍 
Doing a quick review to make sure I didn't miss anything then hitting approve 🚀